### DISCUSSION TITLE
Deploy Nextjs in GCP using serverless in a proper way

### DISCUSSION THREAD

**Discussion by rafinskipg** [vercel/next.js#14353] [Help]
Hello.  I need some help defining a plan to deploy **correctly** in google cloud platform.

The documentation for deployments encourages the use of Vercel. But our client can not approve the use of multi-cloud platforms because of security reasons. We are forced to deploy on google cloud platform.

Is there any way to deploy all the NextJS functions as serverless? And, what other nuances are important in the deployments? 

Our initial idea is this one:
- deploy functions in serverless
- have something to route to those functions, (?)
- where do we deploy the statics? Inside the docker? In a CDN?

I have seen this package that helps to deploy in AWS (but not for google): https://www.npmjs.com/package/serverless-nextjs-plugin

Next has documentation for target serverless but i have no idea what i have to do with it. https://nextjs.org/docs/api-reference/next.config.js/build-target#serverless-target

I am not a devops, if there is anybody who can guide me (and probably other people who need this) to solve this approach, that would be great. 

Basically we need to be able to :
- Set up a CI / CD system targeting serverless in different domains (pro, dev..)

Thanks

**Comment by Tylerian** [comment]
I'm talking from the top of my head but basically you have to configure your Next.js file as below:

```javascript
module.exports = {
  [...]
  assetPrefix: "https://<your-cdn-url>/",
  target: "serverless",
  [...]
}
```

Then, you have to run the commands below in your Dockerfile:
1. Run `next build`
2. Run custom code to upload `.next/static` folder to your CDN
3. Run `next start`

It's not as straightforward as running `vercel --prod` but at least it's doable :)

**Reply by Tylerian** [reply]
Btw I'm assuming you're going to use Google Cloud Run for this. Steps may change a bit if you're using another GCP product.

**Reply by rafinskipg** [reply]
hello @Tylerian  and thanks for your help. That's quite useful. :) I wonder, do you have to delete the old statics from the CDN manually? what about the serverless functions? Where are they deployed?

Thank you!

**Reply by Tylerian** [reply]
Regarding CDN versioning, you can (and btw I think next.js expects you to) version your static folder behind a folder named with your `build_id` (You can find it on `.next/BUILD_ID` file).

Now for the serverless functions, you would have to use AWS Lambda to actually get them working. GCP serverless functions aren't engineered the same way AWS functions are. They aren't thought for frontend (in fact they even forbid you of linking it to a domain name) but backend / sporadic workloads instead. I would go with AWS but if you must use GCP, use Google Cloud Run to power your frontend through serverless technology.

Finally, to set up Next.js serverless in AWS you can have a look at Vercel's deploy runtime [source code](https://github.com/vercel/vercel/tree/master/packages/now-next/src) to find how Vercel does it. In the end, Vercel is based on AWS when running serverless functions.

Btw I've spent quite a bit of time recently to understand how it all works behind the scenes but I'm no GCP nor AWS expert by any means so take my words with a grain of salt.
